### PR TITLE
Collect docs working with x-pack modules

### PR DESCRIPTION
**Pull request by sayden**
More info https://github.com/elastic/beats/issues/10049

Changes necessary in `docs_collector.py` to include modules from `x-pack/metricbeat` folder.

[This commit](https://github.com/sayden/beats/commit/64b26af7ef64f96f2e0eed838829230e4d242713) contains the `*.asciidocs` from `x-pack/metricbeat` folder: aws, aws-ec2, mssql, mssql-db, mssql-performance. Also some file called `modules_list.asciidoc` where the script added some lines. I'm not sure if I have to add this commit with this files too. @ruflin tell me if I have to revert that last commit.

X-Pack files have been generated into `metricbeat/docs` (OSS) which I think it's the expected behavior because afaik the web extracts its content from here.

### CODE REVIEW

File: metricbeat/docs/modules/mssql/db.asciidoc [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+experimental[]
+
+include::../../../module/mssql/db/_meta/docs.asciidoc[]
+
+
+==== Fields
+
+For a description of each field in the metricset, see the
+<<exported-fields-mssql,exported fields>> section.
```
Comment (ruflin) [inline metricbeat/docs/modules/mssql/db.asciidoc]: The key `exported-fields-mssql` does not exist here. Probably the `generate_fields_docs` scripts needs an update here on line 160 to not read the `fields.yml` from the oss beat but from the x-pack version. Perhaps you can open a separate PR for this and rebase this one on top of it. Best also check how Filebeat does it as I assume there it's already working. Perhaps only the path passed to the method must be adjusted.

File: metricbeat/docs/modules/aws/ec2.asciidoc [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+==== Fields
+
+For a description of each field in the metricset, see the
+<<exported-fields-aws,exported fields>> section.
+
+Here is an example document generated by this metricset:
+
+[source,json]
+----
+include::../../../module/aws/ec2/_meta/data.json[]
```
Comment (ruflin) [inline metricbeat/docs/modules/aws/ec2.asciidoc]: These data.json paths must be adjusted to link to the x-pack modules instead.

File: metricbeat/scripts/docs_collector.py [hunk 3]
```diff
@@ -174,18 +185,19 @@ def collect(beat_name):
 
                 metricset_file += "[source,json]\n"
                 metricset_file += "----\n"
-                metricset_file += "include::../../../module/" + module + "/" + metricset + "/_meta/data.json[]\n"
+                metricset_file += "include::../../../module/" + \
```
Comment (ruflin) [inline metricbeat/scripts/docs_collector.py]: Here you need to check if x-pack or not to put in the correct path.

**Comment by ruflin** [comment]
I opened https://github.com/elastic/beats/pull/10236 to solve the fields collection issue: https://github.com/elastic/beats/pull/10236

**Comment by ruflin** [comment]
It seems some paths are still not correct, for example `22:05:12 a2x: ERROR: resource file outside destination directory: ../images/metricbeat-aws-ec2-overview.png`. You can test this locally by running `make docs` in the metricbeat dir.

**Comment by sayden** [comment]
Thanks @ruflin. I was just testing with `collect-docs`

**Comment by sayden** [comment]
I have been testing this and iut seems that the error comes from the path inside the docs file. I mean that by setting the `ec2` doc path to `./images/metricbeat-aws-ec2-overview.png` everything works.

It also seems to be the way to go in modules like Apache as you can see in the last line here.

Should we open a different PR to fix that doc? or should I push it here directly to check?

**Comment by ruflin** [comment]
As these issues are related to the x-pack modules path best push it directly in this PR so we know if it is working directly.

**Comment by ruflin** [comment]
@sayden Now with the transaction_log PR merged this needs an other rebase and `make update` I think.

**Comment by sayden** [comment]
It seems that it was a rebase plus `make collect-docs && make docs`

File: metricbeat/scripts/docs_collector.py [hunk 4]
```diff
@@ -19,20 +21,34 @@ def collect(beat_name):
...
-    for module in sorted(os.listdir(base_dir)):
-
-        module_doc = path + "/" + module + "/_meta/docs.asciidoc"
+    modules_path = [{
+        'base_dir': oss_base_dir,
+        'path': oss_path + "/" + module,
+        'name': module,
+        'metricsets': sorted(os.listdir(oss_path + "/" + module))
+    } for module in filter(lambda module: os.path.isfile(oss_path + "/" + module + "/_meta/docs.asciidoc"),
+                           os.listdir(oss_base_dir))]
```
Comment (jsoriano) [inline metricbeat/scripts/docs_collector.py]: :heart: functional python :slightly_smiling_face:

File: metricbeat/scripts/docs_collector.py [hunk 5]
```diff
@@ -19,20 +21,34 @@ def collect(beat_name):
 
     modules_list = {}
 
-    # Iterate over all modules
-    for module in sorted(os.listdir(base_dir)):
-
-        module_doc = path + "/" + module + "/_meta/docs.asciidoc"
+    modules_path = [{
+        'base_dir': oss_base_dir,
+        'path': oss_path + "/" + module,
```
Comment (jsoriano) [inline metricbeat/scripts/docs_collector.py]: Use `os.path.join` for joining paths?

File: metricbeat/scripts/docs_collector.py [hunk 6]
```diff
@@ -19,20 +21,34 @@ def collect(beat_name):
...
+                           os.listdir(oss_base_dir))]
+
+    if os.path.isdir(os.path.abspath(xpack_base_dir)):
+        modules_path += [{
+            'base_dir': xpack_base_dir,
+            'path': xpack_path + "/" + module,
+            'name': module,
+            'metricsets': sorted(os.listdir(xpack_path + "/" + module)),
+        } for module in filter(lambda module: os.path.isfile(xpack_path + "/" + module + "/_meta/docs.asciidoc"),
+                               os.listdir(xpack_path))]
```
Comment (jsoriano) [inline metricbeat/scripts/docs_collector.py]: The code to generate `modules_path` is twice here and only depends on `X_base_dir` (and `X_path` but can be obtained from the other one). Extract it to a function?

File: metricbeat/scripts/docs_collector.py [hunk 7]
```diff
@@ -19,20 +21,34 @@ def collect(beat_name):
...
+        'metricsets': sorted(os.listdir(oss_path + "/" + module))
+    } for module in filter(lambda module: os.path.isfile(oss_path + "/" + module + "/_meta/docs.asciidoc"),
+                           os.listdir(oss_base_dir))]
+
+    if os.path.isdir(os.path.abspath(xpack_base_dir)):
+        modules_path += [{
+            'base_dir': xpack_base_dir,
+            'path': xpack_path + "/" + module,
+            'name': module,
+            'metricsets': sorted(os.listdir(xpack_path + "/" + module)),
```
Comment (jsoriano) [inline metricbeat/scripts/docs_collector.py]: [Named tuples](https://docs.python.org/2/library/collections.html#collections.namedtuple) are pretty handy in these cases, specially when fields of these objects are accessed so many times.

File: metricbeat/docs/modules/aws/ec2.asciidoc [hunk 8]
```diff
@@ -0,0 +1,23 @@
+////
+This file is generated! See scripts/docs_collector.py
+////
+
+[[metricbeat-metricset-aws-ec2]]
+=== aws ec2 metricset
+
+experimental[]
```
Comment (jsoriano) [inline metricbeat/docs/modules/aws/ec2.asciidoc]: Umm, it appears as experimental but the module is beta, in any case I guess this is not related to this PR //cc @kaiyan-sheng

File: metricbeat/docs/modules/mssql/performance.asciidoc [hunk 9]
```diff
@@ -0,0 +1,23 @@
+////
+This file is generated! See scripts/docs_collector.py
+////
+
+[[metricbeat-metricset-mssql-performance]]
+=== mssql performance metricset
+
+experimental[]
```
Comment (jsoriano) [inline metricbeat/docs/modules/mssql/performance.asciidoc]: Also these metricsets are experimental but the module is beta, is it expected?

File: metricbeat/scripts/docs_collector.py [hunk 10]
```diff
@@ -110,34 +126,33 @@ def collect(beat_name):
...
 
             metricset_file = generated_note
 
             # Add reference to metricset file and include file
             metricset_file += reference + "\n"
 
-            metricset_fields = ""
             with open(metricset_fields_path) as f:
                 metricset_fields = yaml.load(f.read())
                 metricset_fields = metricset_fields[0]
```
Comment (jsoriano) [inline metricbeat/scripts/docs_collector.py]: Amazing scopes in python :unamused:

File: metricbeat/scripts/docs_collector.py [hunk 11]
```diff
@@ -151,21 +166,22 @@ def collect(beat_name):
             if release != "ga":
                 metricset_file += "{}[]\n\n".format(get_release(metricset_fields))
 
-            modules_list[module]["metricsets"][metricset]["release"] = release
+            modules_list[module['name']]["metricsets"][metricset]["release"] = release
```
Comment (jsoriano) [inline metricbeat/scripts/docs_collector.py]: `modules_list[module['name']]` referenced lots of times, it could be created as a single object and assigned to the dict when finished.

**Comment by sayden** [comment]
Thanks for the review @jsoriano :slightly_smiling_face: 

@ruflin I'm not sure if we are in hurry to merge this, do you prefer that I make the changes requested by @jsoriano in a follow up PR?

**Comment by ruflin** [comment]
@sayden If the changes can be made today SGTM but I would prefer to definitively get this in this week.

**Comment by sayden** [comment]
Oh! You were faster @ruflin :smile:

**Comment by sayden** [comment]
I'll merge this today and open an issue to address the changes.

The reason behind this is that I'm not confident to make it work again with so many small changes, slight differences.

I'm also not confident that it's going to be the "final" review (means, no more changes after addressing what is commented here).

**Comment by sayden** [comment]
The follow up issue too https://github.com/elastic/beats/issues/10316

File: metricbeat/docs/modules/aws/ec2.asciidoc (hunk 8 above)
Comment (kaiyan-sheng) [reply metricbeat/docs/modules/aws/ec2.asciidoc]: @jsoriano @sayden Yeah I had a PR changing aws module from experimental to beta. Do we need a rebase here?

Comment (sayden) [reply metricbeat/docs/modules/aws/ec2.asciidoc]: I think (but I'm not sure) there's no need to rebase as the docs generation scripts will do their job once it's merged

**Comment by jsoriano** [comment]
Another option for follow up is to migrate the script to go :innocent:
